### PR TITLE
Gateway Streaming for OpenAI

### PR DESCRIPTION
**Pull request by gabrielfu**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/gabrielfu/mlflow/pull/10520?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/10520/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 10520
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

Implemented streaming support for MLflow AI Gateway with OpenAI provider on chat and completions endpoints.

Example:

```python
from mlflow.gateway import query, set_gateway_uri

set_gateway_uri(gateway_uri="http://127.0.0.1:5000")
route_name = "chat"
data = dict(
    messages=[
        {
            "role": "user",
            "content": "say 'hello world'",
        }
    ],
    candidate_count=1,
    temperature=0.2,
    max_tokens=20,
)


for chunk in query(route_name, data, stream=True):
    print(chunk)
    print("======================================")
```

### How is this PR tested?

- [ ] Existing unit/integration tests
- [X] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [X] Yes. I've updated:
  - [ ] Examples
  - [X] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.

Implemented streaming support for MLflow AI Gateway with OpenAI provider on chat and completions endpoints.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [X] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [X] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by gabrielfu** [comment]
@harupy as discussed, the response follows OpenAI schema. Please let me know if this looks good before I start extending to other providers :)

File: mlflow/gateway/client.py
Comment (harupy) [inline mlflow/gateway/client.py:1]: @gabrielfu Thanks for the PR! We can ignore the client for now. `mlflow.gateway` will be deprecated soon.

File: mlflow/gateway/providers/openai.py [hunk 1]
```diff
@@ -83,13 +99,38 @@ async def chat(self, payload: chat.RequestPayload) -> chat.ResponsePayload:
...
+        async for chunk in stream:
+            chunk = chunk.strip()
+            if not chunk:
+                continue
+
+            data = strip_sse_prefix(chunk.decode("utf-8"))
+            if data == "[DONE]":
+                return
+
+            await asyncio.sleep(0.05)
```
Comment (harupy) [inline mlflow/gateway/providers/openai.py:120]: Can we remove this line?

File: docs/source/llms/gateway/index.rst [hunk 2]
```diff
@@ -899,6 +896,85 @@ The results of the query are:
          }
        }
 
+Streaming
+~~~~~~~~~
+
+Some providers support streaming responses. Streaming responses are useful when you want to
```
Comment (harupy) [inline docs/source/llms/gateway/index.rst:902]: Can we say "The following providers supports ..." instead of `Some providers ...`?

Comment (harupy) [reply docs/source/llms/gateway/index.rst:902]: Never mind this comment.

File: docs/source/llms/gateway/index.rst [hunk 3]
```diff
@@ -899,6 +896,85 @@ The results of the query are:
...
+generated before receiving it. Streaming responses are supported by the following providers:
+
+- OpenAI
+
+To enable streaming responses, set the ``stream`` parameter to ``true`` in your query. For example:
+
+.. code-block:: python
+
+    for chunk in query(route="completions-gpt4", data=data, stream=True):
+        print(chunk)
```
Comment (harupy) [inline docs/source/llms/gateway/index.rst:913]: Can we replace this with `curl` example?

**Comment by harupy** [comment]
@gabrielfu We're working to deprecating `mlflow.gateway` in https://github.com/mlflow/mlflow/pull/10420. This PR contains lots of changes in `mlflow.gateway` and will conflict with this PR. We'll merge it on Friday. Is it possible to rebase after that?

**Comment by gabrielfu** [comment]
> @gabrielfu We're working on deprecating `mlflow.gateway` in #10420. This PR contains lots of changes in `mlflow.gateway` and will conflict with this PR. We'll merge it on Friday. Is it possible to rebase after that?

@harupy no problem, I'll wait for that PR.

File: mlflow/gateway/client.py
Comment (gabrielfu) [reply mlflow/gateway/client.py:1]: Sure!

**Comment by harupy** [comment]
@gabrielfu The gateway-migration has been merged (lots of conflicts as expected...)

**Comment by gabrielfu** [comment]
> @gabrielfu The gateway-migration has been merged (lots of conflicts as expected...)

No problem, let me update this pr :)

**Pull request closed without merging**